### PR TITLE
Fix DateTimeException error classification in varchar casts.

### LINKED ISSUE #5924: DateTimeException misclassified as INTERNAL_ERROR
Seeing the following exception which ends up being classified as an INTERNAL_ERROR instead of a USER_ERROR:

    java.time.DateTimeException: Invalid value for HourOfDay (valid values 0 - 23): 24
    	at java.base/java.time.temporal.ValueRange.checkValidValue(ValueRange.java:311)
    	at java.base/java.time.temporal.ChronoField.checkValidValue(ChronoField.java:717)
    	at java.base/java.time.LocalTime.of(LocalTime.java:339)
    	at java.base/java.time.LocalDateTime.of(LocalDateTime.java:362)
    	at java.base/java.time.ZonedDateTime.of(ZonedDateTime.java:339)
    	at io.prestosql.operator.scalar.timestamp.VarcharToTimestampCast.castToShortTimestamp(VarcharToTimestampCast.java:132)
    	at io.prestosql.operator.scalar.timestamp.VarcharToTimestampCast.castToLegacyShortTimestamp(VarcharToTimestampCast.java:96)
    	at io.prestosql.operator.scalar.timestamp.VarcharToTimestampCast.castToShort(VarcharToTimestampCast.java:57)
    	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:710)
    	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:735)
    	at io.prestosql.sql.InterpretedFunctionInvoker.invoke(InterpretedFunctionInvoker.java:100)
    	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitCast(ExpressionInterpreter.java:1133)
    	at io.prestosql.sql.tree.Cast.accept(Cast.java:91)
    	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
    	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitComparisonExpression(ExpressionInterpreter.java:747)
    	at io.prestosql.sql.tree.ComparisonExpression.accept(ComparisonExpression.java:71)
    	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
    	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitLogicalBinaryExpression(ExpressionInterpreter.java:861)
    	at io.prestosql.sql.tree.LogicalBinaryExpression.accept(LogicalBinaryExpression.java:88)
    	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
    	at io.prestosql.sql.planner.ExpressionInterpreter.optimize(ExpressionInterpreter.java:287)
    	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.rewrite(SimplifyExpressions.java:54)
    	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.lambda$createRewrite$0(SimplifyExpressions.java:79)
    	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$FilterExpressionRewrite.apply(ExpressionRewriteRuleSet.java:226)
    	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$FilterExpressionRewrite.apply(ExpressionRewriteRuleSet.java:207)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.transform(IterativeOptimizer.java:168)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreNode(IterativeOptimizer.java:143)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:108)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
    	at io.prestosql.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:99)
    	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:200)
    	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:189)
    	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:184)
    	at io.prestosql.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:433)
    	at io.prestosql.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:421)
    	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:373)
    	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:232)
    	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:132)
    	at io.prestosql.$gen.Presto_338_12____20201102_161224_2.run(Unknown Source)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    	at java.base/java.lang.Thread.run(Thread.java:834)

### PR DESCRIPTION
**Pull request by johnwhumphreys**
Fixes #5924 - Ensures that varchar casts to timestamps yield user error exceptions with detailed messages when they have improper field values.

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java [hunk 1]
```diff
@@ -53,6 +54,11 @@ public static long castToShort(@LiteralParameter("p") long precision, @SqlType("
         catch (IllegalArgumentException e) {
             throw new PrestoException(INVALID_CAST_ARGUMENT, "Value cannot be cast to timestamp: " + value.toStringUtf8(), e);
         }
+        catch (DateTimeException e) {
+            //Leverage highly specific error message from the source exception.
+            throw new PrestoException(INVALID_CAST_ARGUMENT,
+                    String.format("Value cannot be cast to timestamp; %s. ", e.getMessage()), e);
```
Comment (dain) [inline presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java]: The exception above includes the value.  Do we lose that here?

Comment (johnwhumphreys) [reply presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java]: Good point, it shows you the erroneous part of the value but not the full original value.  Will fix it.

Comment (johnwhumphreys) [reply presto-main/src/main/java/io/prestosql/operator/scalar/timestamp/VarcharToTimestampCast.java]: @dain - Fixed, it now shows the original value as well as the detailed message indicating what was wrong with it.

**Comment by hashhar** [comment]
@johnwhumphreys Can you rebase please? This looks good to be merged but somehow fell through I guess.

**Comment by findepi** [comment]
@johnwhumphreys do you want to rebase?

**Pull request closed without merging**